### PR TITLE
[Merged by Bors] - feat(NumberTheory/Padics): Mahler basis functions

### PR DESCRIPTION
**Pull request by loefflerd**
This PR adds the Mahler basis functions, which are the functions `x ↦ x.choose k` on `ℤ_[p]`. This is preparation for Mahler's theorem (to be added in a subsequent PR), which shows that these functions are a Banach space basis for the space of continuous functions on `ℤ_[p]`.

Partially based on the ETH bachelor thesis of Giulio Caflisch.

 ---

- [x] depends on: #18410

### CODE REVIEW

**Comment by acmepjz** [comment]
Hi! Do you have plan to formalize various distribution spaces over Qp in Colmez's 1998 paper *Theorie d'Iwasawa des representations de de Rham d'un corps local*?

**Comment by loefflerd** [comment]
> Hi! Do you have plan to formalize various distribution spaces over Qp in Colmez's 1998 paper _Theorie d'Iwasawa des representations de de Rham d'un corps local_?

For now, I am concentrating on continuous functions and their dual, measures (tempered distributions of order 0). I believe there is a group at BICMR working on the more general setting of functions and distributions of positive order; there was some [discussion of this on Zulip recently](https://leanprover.zulipchat.com/#narrow/channel/217875-Is-there-code-for-X.3F/topic/Zp.20is.20compact.3F). 

(Mahler's theorem has been formalised before, but has never made it into mathlib; so I am concentrating on getting this theorem in before building more on top of it.)

File: Mathlib/NumberTheory/Padics/MahlerBasis.lean
Comment (ScottCarnahan) [inline Mathlib/NumberTheory/Padics/MahlerBasis.lean:34]: The underlying function is `Ring.choose x k` (from `RingTheory.Binomial`). I'm not sure if you want to go through the trouble of producing a `BinomialRing` instance for `Z_[p]` and proving continuity with this alternative definition, though.  It would however make `mathler_natCast_eq` superfluous.

Comment (loefflerd) [reply Mathlib/NumberTheory/Padics/MahlerBasis.lean:34]: We wondered about using the `BinomialRing` machinery but didn't pursue it, since I wasn't sure how easy it would be to prove continuity from this definition. But if having a BinomialRing instance on Zp is something that would be useful in its own right then we can try that route instead. (I think there are no other examples of BinomialRings at present except N and Z, so having another one can't do any harm.)

Comment (loefflerd) [reply Mathlib/NumberTheory/Padics/MahlerBasis.lean:34]: So I tried, and it wasn't much fun. Obtaining a `BinomialRing` instance on Zp is not too bad, it's just a reorganisation of the code we already have for showing that `‖mahler k‖ = 1`. But if you define `mahler k` using `Ring.choose`, deriving continuity from this seems to be quite awkward; I got as far as proving a goal which printed as `(descPochhammer ℤ k).smeval x = (descPochhammer ℤ k).smeval x`, for `x : ℤ_[p]`, but unfortunately the LHS and RHS seem to be using two different definitions of `smul` which Mathlib is unable to show are the same.

Comment (loefflerd) [reply Mathlib/NumberTheory/Padics/MahlerBasis.lean:34]: See #18410.

**Comment by loefflerd** [comment]
@ScottCarnahan Thanks for the suggestion to try using `BinomialRing`! After some initial struggles this eventually worked out rather nicely. So the PR now proves that Z_p is a binomial ring and derives the results from that, as you suggested.

**Comment by ScottCarnahan** [comment]
This looks great. Thanks for cleaning up my diamond!

**Review by ScottCarnahan — APPROVED** [review]
This looks great. Thanks for cleaning up my diamond!

File: Mathlib/NumberTheory/Padics/MahlerBasis.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Padics/MahlerBasis.lean:39]:
```suggestion
  -- find `n : ℕ` such that `‖f x - f n‖ ≤ ‖k!‖`
```
At least, this is what is actually done here...

File: Mathlib/NumberTheory/Padics/MahlerBasis.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Padics/MahlerBasis.lean:44]: I assume this comment should be modified in the same way.

File: Mathlib/NumberTheory/Padics/MahlerBasis.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Padics/MahlerBasis.lean:46]: You actually use that the binomial coefficient is a natural number.

**Review by MichaelStollBayreuth — APPROVED** [review]
LGTM, modulo the suggested fixes to comments in the code.

File: Mathlib/NumberTheory/Padics/MahlerBasis.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Padics/MahlerBasis.lean:64]:
```suggestion
@[fun_prop]
lemma continuous_multichoose (k : ℕ) : Continuous (fun x : ℤ_[p] ↦ Ring.multichoose x k) := by
  simp only [Ring.multichoose, BinomialRing.multichoose, continuous_induced_rng]
  fun_prop
```

File: Mathlib/NumberTheory/Padics/MahlerBasis.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Padics/MahlerBasis.lean:68]:
```suggestion
@[fun_prop]
lemma continuous_choose (k : ℕ) : Continuous (fun x : ℤ_[p] ↦ Ring.choose x k) := by
  simp only [Ring.choose]
  fun_prop
```

**Comment by loefflerd** [comment]
Thanks Michael! All comments now dealt with.

**Comment by MichaelStollBayreuth** [comment]
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
